phoenix-core/src/main/java/org/apache/phoenix/filter/SystemCatalogViewIndexIdFilter.java

```diff
@@ -0,0 +1,160 @@
+import static org.apache.phoenix.util.ViewIndexIdRetrieveUtil.VIEW_INDEX_ID_SMALLINT_TYPE_VALUE_LEN;
+
+public class SystemCatalogViewIndexIdFilter extends FilterBase implements Writable {
+    private int clientVersion;
+
+    public SystemCatalogViewIndexIdFilter() {
+    }
+
+    public SystemCatalogViewIndexIdFilter(int clientVersion) {
+        this.clientVersion = clientVersion;
+    }
+
+    @Override
+    public ReturnCode filterKeyValue(Cell keyValue) {
+        return ReturnCode.INCLUDE_AND_NEXT_COL;
+    }
+
+    @Override
+    public boolean hasFilterRow() {
+        return true;
+    }
+
+    @Override
+    public void filterRowCells(List<Cell> kvs) throws IOException {
+        Cell viewIndexIdCell = KeyValueUtil.getColumnLatest(
+                GenericKeyValueBuilder.INSTANCE, kvs,
+                DEFAULT_COLUMN_FAMILY_BYTES, VIEW_INDEX_ID_BYTES);
+
+        /*
+            We retrieve the VIEW_INDEX_ID cell from SMALLINT to BIGINT or BIGINT to SMALLINT if and
+            only if VIEW_INDEX_ID is included as part of the projected column.
+
+            This is combination of diff client created view index looks like:
+
+            client                  VIEW_INDEX_ID(Cell number of bytes)     VIEW_INDEX_ID_DATA_TYPE
+        pre-4.14                        2 bytes                                     NULL
+        post-4.15[config smallint]      2 bytes                                     5(smallint)
+        post-4.15[config bigint]        8 bytes                                     -5(bigint)
+         */
+        if (viewIndexIdCell != null) {
+            int type = NULL_DATA_TYPE_VALUE;
+            Cell viewIndexIdDataTypeCell = KeyValueUtil.getColumnLatest(
+                    GenericKeyValueBuilder.INSTANCE, kvs,
+                    DEFAULT_COLUMN_FAMILY_BYTES, VIEW_INDEX_ID_DATA_TYPE_BYTES);
+            if (viewIndexIdDataTypeCell != null) {
+                type = (Integer) PInteger.INSTANCE.toObject(
+                        viewIndexIdDataTypeCell.getValueArray(),
+                        viewIndexIdDataTypeCell.getValueOffset(),
+                        viewIndexIdDataTypeCell.getValueLength(),
+                        PInteger.INSTANCE,
+                        SortOrder.ASC);
+            }
+            if (this.clientVersion < MIN_SPLITTABLE_SYSTEM_CATALOG) {
+                /*
+                    For pre-4.15 client select query cannot include VIEW_INDEX_ID_DATA_TYPE as part
+                    of the projected columns; for this reason, the TYPE will always be NULL. Since
+                    the pre-4.15 client always assume the VIEW_INDEX_ID column is type of SMALLINT,
+                    we need to retrieve the BIGINT cell to SMALLINT cell.
+
+                    VIEW_INDEX_ID_DATA_TYPE,      VIEW_INDEX_ID(Cell representation of the data)
+                        NULL,                         SMALLINT         -> DO NOT RETRIEVE
```

Comment:
By DO NOT RETRIEVE, you mean DO NOT CONVERT right? Can you please clarify this in comments?